OpenRA.Mods.Common/Traits/Air/Aircraft.cs

```diff
@@ -1088,6 +1089,109 @@ public void ResolveOrder(Actor self, Order order)
 
 		#endregion
 
+		bool CanOrder(Actor actor, CPos cell)
+		{
+			if (actor == null || actor.IsDead)
+				return false;
+
+			var aircraft = actor.TraitOrDefault<Aircraft>();
+			return aircraft != null && (aircraft.Info.MoveIntoShroud || self.Owner.Shroud.IsExplored(cell));
+		}
+
+		public IEnumerable<Pair<Actor, CPos>> GroupMove(CPos cell, Order order)
+		{
+			// Make sure we place the units that are closer to the target first to prevent unnecessary shuffles.
+			var selection = order.ExtraActors.Where(a => CanOrder(a, cell)).OrderBy(a => (a.Location - cell).LengthSquared);
+			var n = selection.Count();
+
+			// Prevent division by zero errors.
+			if (n < 1)
+				yield break;
+
+			// Calculate average group location
+			var center = CVec.Zero;
+			foreach (var a in selection)
+				center += (a.Location - CPos.Zero);
+
+			center /= n;
+
+			// Calculate variance of unit locations
+			var variance = 0;
+			foreach (var a in selection)
+				variance += CVec.Dot((a.Location - CPos.Zero) - center, (a.Location - CPos.Zero) - center);
+
+			variance /= n;
+
+			// Calculate minimum formation radius.
+			var minRadiusSq = 0;
+			foreach (var a in selection)
+				minRadiusSq += (int)a.Trait<Aircraft>().Info.IdealSeparation.LengthSquared;
+
+			minRadiusSq /= (1024 * 1024 * 2);
+
+			// Determine individual destination cells
+			var claimed = new List<CPos> { };
+			foreach (var a in selection)
+			{
+				var aircraft = a.Trait<Aircraft>();
+				var tile = aircraft.GetDestinationCell(cell, center, variance, minRadiusSq, claimed);
+
+				if (!order.Queued)
+					aircraft.UnReserve();
+
+				yield return Pair.New(a, tile);
+			}
+		}
+
+		public CPos GetDestinationCell(CPos targetCell, CVec center, int variance, int minRadiusSq, List<CPos> claimed)
```

Comment:
Could you please add a comment outlining the logic behind the maths and magic constants (3/2 in the variance, 6-cell limit for unclaimed cells, etc) in these methods?

Reply:
The 6-cell limit is not something I want to leave in. Conceptually, there should be no arbitrary limit, but I'm not sure how to do that in a performant way.

I forgot how exactly I arrived at the 3/2. I think I started with calculating what the variance would be for the minimal circular formation and then that didn't get the right behaviour in practice, so I just used trial and error to get to this value.